src/sing/domain.ts

```diff
@@ -730,16 +734,132 @@ export function applyPitchEdit(
   );
   const phraseQueryEndFrame = phraseQueryStartFrame + phraseQueryFrameLength;
 
-  // ピッチ編集をf0に適用する
-  const startFrame = Math.max(0, phraseQueryStartFrame);
-  const endFrame = Math.min(pitchEditData.length, phraseQueryEndFrame);
-  for (let i = startFrame; i < endFrame; i++) {
-    const phoneme = framePhonemes[i - phraseQueryStartFrame];
-    const voiced = !unvoicedPhonemes.includes(phoneme);
-    if (voiced && pitchEditData[i] !== VALUE_INDICATING_NO_DATA) {
-      f0[i - phraseQueryStartFrame] = pitchEditData[i];
+  // 処理対象のフレーム範囲を計算（負のフレームは0にクリップ）
+  const processStartFrame = Math.max(0, phraseQueryStartFrame);
+  const processEndFrame = phraseQueryEndFrame;
+
+  const logWithMinOne = (value: number) => Math.log(Math.max(1, value));
+
+  const frameInfos: {
+    isEdited: boolean;
+    isVoiced: boolean;
+    indexInTrack: number;
+  }[] = [];
+  const logF0 = f0.map(logWithMinOne);
+
+  // 元の（推論された）ピッチとの差分を格納する配列
+  const logF0Diff: number[] = [];
+
+  // 各フレームの情報を収集し、対数f0の差分を計算する
+  for (let i = processStartFrame; i < processEndFrame; i++) {
+    const indexInPhrase = i - phraseQueryStartFrame;
+    const phoneme = framePhonemes[indexInPhrase];
+    const isVoiced = !UNVOICED_PHONEMES.includes(phoneme);
+
+    // NOTE: 無声区間またはpitchEditDataの範囲外の場合はVALUE_INDICATING_NO_DATAを使用
+    let editValue = VALUE_INDICATING_NO_DATA;
+    if (isVoiced && i < pitchEditData.length) {
+      editValue = pitchEditData[i];
     }
+
+    const isEdited = editValue !== VALUE_INDICATING_NO_DATA;
+
+    frameInfos.push({
+      isEdited,
+      isVoiced,
+      indexInTrack: i,
+    });
+
+    if (isEdited) {
+      const originalLogF0 = logF0[indexInPhrase];
+      const editedLogF0 = logWithMinOne(editValue);
+
+      logF0Diff.push(editedLogF0 - originalLogF0);
+    } else {
+      logF0Diff.push(0);
+    }
+  }
+
+  // 編集データが全くない場合は何もしない
+  const hasEditData = frameInfos.some((frameInfo) => frameInfo.isEdited);
+  if (!hasEditData) {
+    return;
   }
+
+  // 編集/未編集の境界（ジャンプポイント）を検出し、遷移長を計算する
```

Comment:
`/`は意味が人によって異なりうるので危ないかも（編集÷未編集の境界、と取れるかもしれない）
細かいけど基本記号は使わないのが安全かな～と
```suggestion
  // 編集と未編集の境界（ジャンプポイント）を検出し、遷移長を計算する
```